codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java

```diff
@@ -506,57 +517,79 @@ public String integerShape(IntegerShape shape) {
       );
     } else {
       return """
-      func() int32 {
+      return func() int32 {
           var b = %s.(int32)
           return b
       }()""".formatted(dataSource);
     }
   }
 
   @Override
-  public String longShape(LongShape shape) {
+  public String longShape(final LongShape shape) {
     writer.addImportFromModule("github.com/dafny-lang/DafnyRuntimeGo", "dafny");
-    return (
-      """
-      func() *int64 {
-          var b int64
-          if %s == nil {
-              return nil
-          }
-          b = %s.(int64)
-          return &b
-      }()"""
-    ).formatted(dataSource, dataSource);
+    if (isOptional) {
+      return (
+        """
+        return func() *int64 {
+            var b int64
+            if %s == nil {
+                return nil
+            }
+            b = %s.(int64)
+            return &b
+        }()"""
+      ).formatted(dataSource, dataSource);
+    } else {
+      return """
+      return func() int64 {
+          var b = %s.(int64)
+          return b
+      }()
+          """.formatted(dataSource);
+    }
   }
 
   @Override
-  public String doubleShape(DoubleShape shape) {
+  public String doubleShape(final DoubleShape shape) {
     writer.addImportFromModule("github.com/dafny-lang/DafnyRuntimeGo", "dafny");
     writer.addUseImports(SmithyGoDependency.MATH);
-    return """
-    func () *float64 {
-        var b []byte
-    if %s == nil {
-        return nil
+    if (isOptional) {
+      return """
+      return func () *float64 {
```

Comment:
Use the if check with %s to figure out if it's a pointer or not.

Reply:
Let's do this as a clean up task at the end. Every non container shape is written in this way. I will add a TODO.